forge/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -0,0 +1,300 @@
+            .or_else(|| util::apply(self, data, &decoded))
+            .or_else(|| expect::apply(self, data, &decoded))
+            .or_else(|| fuzz::apply(data, &decoded))
+            .or_else(|| ext::apply(self.ffi, &decoded))
+            .ok_or_else(|| "Cheatcode was unhandled. This is a bug.".to_string().encode())?
+    }
+}
+
+impl<DB> Inspector<DB> for Cheatcodes
+where
+    DB: Database,
+{
+    fn call(
+        &mut self,
+        data: &mut EVMData<'_, DB>,
+        call: &CallInputs,
+        _: bool,
+    ) -> (Return, Gas, Bytes) {
+        if call.contract == *CHEATCODE_ADDRESS {
+            match self.apply_cheatcode(data, call.context.caller, call) {
+                Ok(retdata) => (Return::Return, Gas::new(0), retdata),
+                Err(err) => (Return::Revert, Gas::new(0), err),
+            }
+        } else {
+            // Handle expected calls
+            if let Some(expecteds) = self.expected_calls.get_mut(&call.contract) {
+                if let Some(found_match) = expecteds.iter().position(|expected| {
+                    expected.len() <= call.input.len() && expected == &call.input[..expected.len()]
+                }) {
+                    expecteds.remove(found_match);
+                }
+            }
+
+            // Handle mocked calls
+            if let Some(mocks) = self.mocked_calls.get(&call.contract) {
+                if let Some(mock_retdata) = mocks.get(&call.input) {
+                    return (Return::Return, Gas::new(0), mock_retdata.clone())
+                } else if let Some((_, mock_retdata)) =
+                    mocks.iter().find(|(mock, _)| *mock == &call.input[..mock.len()])
+                {
+                    return (Return::Return, Gas::new(0), mock_retdata.clone())
+                }
+            }
+
+            (Return::Continue, Gas::new(0), Bytes::new())
+        }
+    }
+
+    fn initialize_interp(
+        &mut self,
+        interpreter: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        _: bool,
+    ) -> Return {
+        // Apply our prank
+        if let Some(prank) = &self.prank {
+            if data.subroutine.depth() >= prank.depth &&
+                interpreter.contract.caller == prank.prank_caller
+            {
+                // At the target depth we set `msg.sender`
+                if data.subroutine.depth() == prank.depth {
+                    interpreter.contract.caller = prank.new_caller;
+                }
+
+                // At the target depth, or deeper, we set `tx.origin`
+                if let Some(new_origin) = prank.new_origin {
+                    data.env.tx.caller = new_origin;
+                }
+            }
+        }
+
+        Return::Continue
+    }
+
+    fn step(&mut self, interpreter: &mut Interpreter, _: &mut EVMData<'_, DB>, _: bool) -> Return {
+        // Record writes and reads if `record` has been called
+        if let Some(storage_accesses) = &mut self.accesses {
```

Comment:
@brockelmore we should eventually use this slot read/write trace logic for determining the parallel evm dag